File: aac/agentsKeeper.py
#import sys
#import json
#from lxml import etree
#import os
#import time
#import re
import sqlite3


#--------------------------------------------------------------------------------------------------------------

_logName = "agentsKeeper"
import logging
logger = logging.getLogger(_logName)

#--------------------------------------------------------------------------------------------------------------

class agentsDataKeeper:

    #-------------
    def __init__(self, data_folder_name):
        self._db_file_name = data_folder_name + "/agents.db"
        self._connection = None

    #-------------
    def __del__(self):
        if not self._connection is None:
            self._connection.commit()
            self._connection.close()

    #-------------
    def init_data(self):

        if self._connection is None:
            try:
                self._connection = sqlite3.connect(self._db_file_name)
            except sqlite3.Error as e:
                logger.error(f"Error connecting to database {repr(self._db_file_name)}: {e}")
                raise

        self._create_tables_if_needed()

    #-------------

    def _table_exists(self, table_name, cursor):
        cursor.execute("SELECT name FROM sqlite_master WHERE type='table' AND name=?", (table_name,) )
        return not cursor.fetchone() is None 

    #-------------

    def _create_tables_if_needed(self):
        cursor = self._connection.cursor()

        for tname in ('Agents','Tags'):
            logger.info(f"Table {repr(tname)} in database {repr(self._db_file_name)} {'exists' if self._table_exists(tname, cursor) else 'to be created'}")

        cursor.execute( '''
            CREATE TABLE IF NOT EXISTS Agents (
                agent_id TEXT PRIMARY KEY,
                branch TEXT,
                descr TEXT,
                location TEXT,
                extra TEXT
            )
        ''')

        cursor.execute( '''
            CREATE TABLE IF NOT EXISTS Tags (
                tag_id INTEGER PRIMARY KEY AUTOINCREMENT,
                agent_id TEXT,
                tag TEXT,
                FOREIGN KEY (agent_id) REFERENCES Agents (agent_id)
            )
        ''')

        self._connection.commit()
        cursor.close()

    #-------------

    def get_all_agent_ids(self):
        cursor = self._connection.cursor()
        cursor.execute("SELECT agent_id FROM Agents")
        ret = [ row[0] for row in iter(cursor.fetchone,None) ]
        cursor.close()
        return ret

    #-------------

    def get_branch_name(self, agent_id):
        cursor = self._connection.cursor()
        cursor.execute("SELECT branch FROM Agents WHERE agent_id=?", (agent_id,))
        row = cursor.fetchone()
        cursor.close()
        return None if row is None else row[0]

    #-------------

    def get_agent_dict(self, agent_id, with_tags=False):
        cursor = self._connection.cursor()

        cursor.execute("SELECT * FROM Agents WHERE agent_id=?", (agent_id,))
        row = cursor.fetchone()
        ret = None if row is None else dict(zip( ("agent_id","branch","descr","location","extra"), row ))

        if with_tags and not ret is None:
            cursor.execute("SELECT tag FROM Tags WHERE agent_id=?", (agent_id,))
            ret['tags'] = [row[0] for row in cursor.fetchall()]

        cursor.close()
        return ret
        
    #-------------
    '''
    def change_agent_branch(self, agent_id, new_branch_name):
        cursor = self._connection.cursor()
        cursor.execute("UPDATE Agents SET branch=? WHERE agent_id=?", (new_branch_name, agent_id,))
        ret = cursor.rowcount > 0
        self._connection.commit()
        cursor.close()
        return ret
    '''
    #-------------

    def add_agent(self, agent_id, branch, descr="", location="", extra="", tags=()):
        cursor = self._connection.cursor()
        cursor.execute( "INSERT INTO Agents (agent_id, branch, descr, location, extra) VALUES (?, ?, ?, ?, ?)", (agent_id, branch, descr, location, extra,) )
        for tag in tags:
            cursor.execute("INSERT INTO Tags (agent_id, tag) VALUES (?, ?)", (agent_id, tag,))
        self._connection.commit()
        cursor.close()

    #-------------

    def delete_agent(self, agent_id):
        cursor = self._connection.cursor()

        cursor.execute("DELETE FROM Tags WHERE agent_id=?", (agent_id,))

        cursor.execute("DELETE FROM Agents WHERE agent_id=?", (agent_id,))
        deleted_rows = cursor.rowcount

        self._connection.commit()
        cursor.close()

        return deleted_rows>0

    #-------------

    def get_agents_by_branch_list(self, branch_names):
        cursor = self._connection.cursor()

        query = "SELECT agent_id,branch FROM Agents WHERE branch IN ({})".format(",".join(["?"] * len(branch_names)))
        cursor.execute(query, branch_names)
        agents = cursor.fetchall()

        cursor.close()
        return agents


#--------------------------------------------------------------------------------------------------------------
# Some self-test functionality

if __name__ == '__main__':

    import configTestLogging
    configTestLogging.config(_logName)

    #~~~~~~~~~~~~~~~~~~~~~~~

    def main():        
        test = agentsDataKeeper("DATA")
        test.init_data()

    main()

File: golang/agentskeeper.go
package main

import (
    "database/sql"
    "fmt"
    "sort"
    "strings"
    _ "modernc.org/sqlite"
)

type agentsKeeper struct {
    dbFile string
    db     *sql.DB
}

func newAgentsKeeper(dataFolder string) *agentsKeeper {
    return &agentsKeeper{dbFile: dataFolder + "/agents.db"}
}

func (ak *agentsKeeper) initData() error {
    if ak.db != nil {
        return nil
    }

    db, err := sql.Open("sqlite", ak.dbFile)
    if err != nil {
        return err
    }

    ak.db = db
    return ak.createTablesIfNeeded()
}

func (ak *agentsKeeper) createTablesIfNeeded() error {
    _, err := ak.db.Exec(`
        CREATE TABLE IF NOT EXISTS Agents (
            agent_id TEXT PRIMARY KEY,
            branch TEXT,
            descr TEXT,
            location TEXT,
            extra TEXT
        )
    `)
    if err != nil {
        return err
    }

    _, err = ak.db.Exec(`
        CREATE TABLE IF NOT EXISTS Tags (
            tag_id INTEGER PRIMARY KEY AUTOINCREMENT,
            agent_id TEXT,
            tag TEXT,
            FOREIGN KEY (agent_id) REFERENCES Agents (agent_id)
        )
    `)
    return err
}

func (ak *agentsKeeper) close() {
    if ak.db != nil {
        _ = ak.db.Close()
        ak.db = nil
    }
}

func (ak *agentsKeeper) getAllAgentIds() []string {
    if ak.db == nil {
        return []string{}
    }
    rows, err := ak.db.Query(`SELECT agent_id FROM Agents`)
    if err != nil {
        return []string{}
    }
    defer rows.Close()

    out := []string{}
    for rows.Next() {
        var agentID string
        if err := rows.Scan(&agentID); err == nil {
            out = append(out, agentID)
        }
    }
    sort.Strings(out)
    return out
}

func (ak *agentsKeeper) getBranchName(agentID string) (string, bool) {
    if ak.db == nil {
        return "", false
    }
    row := ak.db.QueryRow(`SELECT branch FROM Agents WHERE agent_id = ?`, agentID)
    var branch string
    err := row.Scan(&branch)
    if err == sql.ErrNoRows {
        return "", false
    }
    if err != nil {
        return "", false
    }
    return branch, true
}

func (ak *agentsKeeper) getAgentDict(agentID string, withTags bool) map[string]interface{} {
    if ak.db == nil {
        return nil
    }
    row := ak.db.QueryRow(`SELECT agent_id, branch, descr, location, extra FROM Agents WHERE agent_id = ?`, agentID)
    var descr, location, branch, extra string
    if err := row.Scan(&agentID, &branch, &descr, &location, &extra); err != nil {
        return nil
    }

    ret := map[string]interface{}{
        "agent_id": agentID,
        "branch":   branch,
        "descr":    descr,
        "location": location,
        "extra":    extra,
    }

    if withTags {
        rows, err := ak.db.Query(`SELECT tag FROM Tags WHERE agent_id = ?`, agentID)
        if err == nil {
            defer rows.Close()
            tags := []string{}
            for rows.Next() {
                var t string
                if err := rows.Scan(&t); err == nil {
                    tags = append(tags, t)
                }
            }
            ret["tags"] = tags
        }
    }
    return ret
}

func (ak *agentsKeeper) addAgent(agentID, branch, descr, location, extra string, tags []string) error {
    if ak.db == nil {
        return fmt.Errorf("database is not initialized")
    }
    tx, err := ak.db.Begin()
    if err != nil {
        return err
    }
    if _, err := tx.Exec(`INSERT INTO Agents (agent_id, branch, descr, location, extra) VALUES (?, ?, ?, ?, ?)`, agentID, branch, descr, location, extra); err != nil {
        _ = tx.Rollback()
        return err
    }
    for _, t := range tags {
        if t == "" {
            continue
        }
        if _, err := tx.Exec(`INSERT INTO Tags (agent_id, tag) VALUES (?, ?)`, agentID, t); err != nil {
            _ = tx.Rollback()
            return err
        }
    }
    return tx.Commit()
}

func (ak *agentsKeeper) deleteAgent(agentID string) error {
    if ak.db == nil {
        return fmt.Errorf("database not initialized")
    }
    tx, err := ak.db.Begin()
    if err != nil {
        return err
    }
    if _, err := tx.Exec(`DELETE FROM Tags WHERE agent_id = ?`, agentID); err != nil {
        _ = tx.Rollback()
        return err
    }
    res, err := tx.Exec(`DELETE FROM Agents WHERE agent_id = ?`, agentID)
    if err != nil {
        _ = tx.Rollback()
        return err
    }
    affected, _ := res.RowsAffected()
    if err := tx.Commit(); err != nil {
        return err
    }
    if affected == 0 {
        return fmt.Errorf("agent not found")
    }
    return nil
}

func (ak *agentsKeeper) getAgentsByBranches(branchNames []string) [][2]string {
    if ak.db == nil || len(branchNames) == 0 {
        return [][2]string{}
    }

    placeholders := make([]string, len(branchNames))
    args := make([]interface{}, len(branchNames))
    for i, b := range branchNames {
        placeholders[i] = "?"
        args[i] = b
    }
    q := "SELECT agent_id, branch FROM Agents WHERE branch IN (" + strings.Join(placeholders, ",") + ")"
    rows, err := ak.db.Query(q, args...)
    if err != nil {
        return [][2]string{}
    }
    defer rows.Close()

    out := make([][2]string, 0)
    for rows.Next() {
        var agentID, branch string
        if err := rows.Scan(&agentID, &branch); err == nil {
            out = append(out, [2]string{agentID, branch})
        }
    }
    return out
}
